cpp/src_prims/matrix/batched_matrix.h

```diff
@@ -0,0 +1,465 @@
+#pragma once
+
+#include <functional>
+#include <stdexcept>
+#include <tuple>
+#include <unordered_map>
+#include <vector>
+
+#include <linalg/binary_op.h>
+#include <linalg/cublas_wrappers.h>
+#include <memory>
+
+#include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+
+namespace MLCommon {
+namespace Matrix {
+
+std::shared_ptr<double*> init(double* A, std::pair<int, int> shape,
+                              int num_batches, bool gpu = true);
+
+__global__ void identity_matrix_kernel(double** I, int r);
+
+template <typename T>
+void cudaFreeT(T* ptr) {
+  CUDA_CHECK(cudaFree(ptr));
+}
+
+void BMM_Allocate(std::pair<int, int> shape, int num_batches, double*& A_dense,
+                  double**& A_array, bool setZero) {
+  int m = shape.first;
+  int n = shape.second;
+
+  allocate(A_dense, m * n * num_batches, setZero);
```

Comment:
please do not use the `allocate` method. Either accept it as an input or use the `deviceAllocator` and `device_buffer` to manage them internally.